weed/shell/command_volume_fsck.go

```diff
@@ -582,10 +563,25 @@ func (c *commandVolumeFsck) oneVolumeFileIdsSubtractFilerFileIds(dataNodeId stri
 	}
 
 	var orphanFileCount uint64
-	if err = volumeFileIdDb.AscendingVisit(func(n needle_map.NeedleValue) error {
+	cutoffFrom := c.collectCutoffFrom.Add(-*c.cutoffTimeAgo)
+	cutoffFromAtNs := uint64(cutoffFrom.UnixNano())
+	doCutoffOfLastNeedle := true
+	if err = volumeFileIdDb.DescendingVisit(func(n needle_map.NeedleValue) error {
```

Comment:
this is descending based on needle id. But
1. The needle ids are not always monotonically increasing, e.g. for snow flake needle id generation.
2. The needle could get updated or deleted. The largest needle id is not necessarily the last appended entry.

Reply:
> 1. The needle ids are not always monotonically increasing, e.g. for snow flake needle id generation.
> 2. The needle could get updated or deleted. The largest needle id is not necessarily the last appended entry.

In this case, both options are bad, since it is necessary to make a grpc request for all needles

Reply:
the binary search will do this efficiently, since the AppendAtNs is monotonically increasing.

Reply:
> the binary search will do this efficiently, since the AppendAtNs is monotonically increasing.

In this case, the Binary search algorithm is not suitable because algorithm for searching for an object by a given attribute in a set of objects ordered by its own attribute

You yourself wrote that objects are not sorted by AppendAtNs or even by needle id

Reply:
> You yourself wrote that objects are not sorted by AppendAtNs or even by needle id

objects are not sorted by modified time or needle id. but they are sorted by AppendAtNs.
AppendAtNs is for appending time to the .dat file. Modified time are set from the client side.

Reply:
@chrislusf I have two ideas:
1.  Cut-off time to send to a remote call to bulk remove of needles. And on the side by the will of the volume server to check already AppendAtNs.
2.  Binary search can be used after getting missing needles (transfer the filtering function from map to reduce)